Add general option -output-directory

herbelin opened the pull request:
Depends on, and refines #16126 (which added option `Set Extraction Output Directory`).

This PR adds an option `-output-directory` (with word `directory` in full to be consistent with using a full word in option `Set Extraction Output Directory` in #16126, and more generally with the general use of non-abbreviated words in commands) with synonym `-output-dir` (to be consistent with other command line options such as `-native-output-dir`).

This option is used by default for extraction (when `Extraction Output Directory` is unset), `Redirect "file"`, and `Print Universes "file"` when `"file"` is not an absolute path.

Together with #16126, this should provide:
- an indirect answer to #8649 which asks for a way to `Redirect` relatively to the repository where the `.vo` file is output
- to remove the command `Cd` #16119 (or to replace it by an option `Set Output Directory` interactively mimicking `-output-directory`, though this might not be what most developers would recommend) 

Closes #8649
+ #16126 closing #9148

- [ ] Added / updated **test-suite**.
- [x] Added **changelog**.
- [x] Added / updated **documentation**.

Open questions:
- Mentioned in #8052 (dune) but I'm not specialist of dune.
- Can it have an effect on #4953 (validity of filenames for `Redirect`/`Extraction`)?
- Should we have also more specific options `-redirect-output-directory`, `-extraction-output-directory`?

Alizter commented:
Do we want to make the option agnostic to extraction ?

herbelin commented:
> Do we want to make the option agnostic to extraction ?

What do you mean? Isn't it already independent of extraction per se?

Alizter commented:
@herbelin right but the option is still named extraction output directory?

herbelin commented:
With this PR, there are two levels of options, one on the command line `-output-directory dir` (for all commands with an output) and one using `Set` (specifically for extraction, and taking precedence over `-output-directory dir` if the former is already set).

Is it confusing? If yes, we may also add, say `Set Redirect Output Directory`. Or having only `-output-directory dir` but then, that would not solve the CoqIDE MacOS app problem mentioned in #14721.

Alizter commented:
OK then I was just confused.

herbelin commented:
Same as #16126, was there a reason to keep this PR draft? What to expect more from it? It is unclear to me whether there are alternative proposals around?

jfehrle commented on doc/sphinx/practical-tools/coq-commands.rst:
```diff
@@ -365,6 +365,9 @@ and ``coqtop``, unless stated otherwise:
 
 :-native-output-dir *dir*: Set the directory in which to put the aforementioned
   ``.cmxs`` for :tacn:`native_compute`. Defaults to ``.coq-native``.
+:-output-directory *dir*, -output-dir *dir*: Set the directory in which some commands,
+  such as :ref:`extraction` commands, or :cmd:`Redirect`, or :cmd:`Print Universes`,
+  may decide to output the files they create.
```
```suggestion
:-output-directory *dir*, -output-dir *dir*: Sets the output directory for commands that
  write output to files, such as :ref:`extraction` commands, :cmd:`Redirect` and :cmd:`Print Universes`.
```

"may decide" seems too vague.  It's apt to be frustrating for the user if they use the option with a command that doesn't support it.  Can you be more precise?  Better yet, how about making sure that all relevant commands support the option?

herbelin replied:
>  Better yet, how about making sure that all relevant commands support the option?

At the time of writing the PR, all commands that print were taken into account. I don't know how to ensure this in the future though.

> "may decide" seems too vague

The "may decide" refers to extraction which has an option to override the `-output-directory` command. Your formulation is however ok to me, so adopting.

jfehrle commented on doc/sphinx/proof-engine/vernacular-commands.rst:
```diff
@@ -865,6 +865,9 @@ Quitting and debugging
    Executes :n:`@sentence`, redirecting its
    output to the file ":n:`@string`.out".
 
+   The directory in which the file is created (or overriden) can be set
+   with the command line option `-output-directory` (see
+   :ref:`command-line-options`).
```
See my comment above in universe-polymorphism.rst.

herbelin replied:
Modified too, thanks.

jfehrle reviewed (commented):
Wording suggestions

herbelin commented on doc/sphinx/addendum/universe-polymorphism.rst:
```diff
@@ -608,6 +608,9 @@ Printing universes
    If :n:`@string` ends in ``.dot`` or ``.gv``, the constraints are printed in the DOT
    language, and can be processed by Graphviz tools. The format is
    unspecified if `string` doesn’t end in ``.dot`` or ``.gv``.
+   The directory in which the file is created (or overriden) can be set
+   with the command line option `-output-directory` (see
+   :ref:`command-line-options`).
```
If `@string` is absolute, it is used without looking at `-output-directory`. If `@string` is relative, even if including `/`, it is relative to the output directory. I clarified it.

herbelin commented:
Doc updated + fix of `Print Universes "string"`.

herbelin commented:
@coqbot run full ci

herbelin commented:
@skyskimmer: would you agree to assign this PR?

herbelin commented on vernac/topfmt.ml:
```diff
@@ -400,8 +400,22 @@ let print_err_exn any =
+    match Filename.is_relative fname, !Flags.output_directory with
+    | false, _ | _, None ->
+      let pwd = Sys.getcwd () in warn_using_current_directory pwd;
+      Filename.concat pwd fname
```
Yes

herbelin commented on vernac/vernacentries.ml:
```diff
@@ -317,6 +317,12 @@ let print_registered () =
   hov 0 (prlist_with_sep fnl pr_lib_ref @@ Coqlib.get_lib_refs ())
 
+let warn_using_current_directory =
+  CWarnings.create ~name:"print-universes-default-directory" ~category:CWarnings.CoreCategories.filesystem
```
Ok, done. I used a common warning name then ("default-output-directory").

herbelin commented on plugins/extraction/table.ml:
```diff
@@ -512,13 +512,12 @@ let { Goptions.get = output_directory } =
-  | None ->
+  | None, None ->
     let pwd = Sys.getcwd () in
     warn_using_current_directory pwd;
```
Which warning are you talking about?

SkySkimmer replied:
The extraction warning. Currently it says "Use Set Extraction Output Directory", maybe it should say "Use Set Extraction Output Directory or -output-directory".

herbelin reviewed (commented):
OK, done.

Note: I got twice the warning with `Redirect`.  Maybe due to the synterp/interp phases, that is that `Redirect` would be active on both phases (??).

SkySkimmer commented on lib/system.ml:
```diff
@@ -225,6 +225,20 @@ let is_in_system_path filename =
+           strbrk "Use command line option \"-output-directory to set a default directory."))
+
+let get_output_path filename =
+  match Filename.is_relative filename, !Flags.output_directory with
+  | false, _ -> filename
+  | true, None ->
+    let pwd = Sys.getcwd () in warn_using_current_directory pwd;
+    Filename.concat pwd filename
+  | true, Some dir -> Filename.concat dir filename
```
```suggestion
  if not (Filename.is_relative filename) then filename
  else match !Flags.output_directory with
  | None ->
    let pwd = Sys.getcwd () in
    warn_using_current_directory pwd;
    Filename.concat pwd filename
  | Some dir -> Filename.concat dir filename
```

SkySkimmer commented on lib/system.mli:
```diff
@@ -58,6 +58,11 @@ val is_in_system_path : string -> bool
 val where_in_path :
   ?warn:bool -> CUnix.load_path -> string -> CUnix.physical_path * string
 
+(** [get_output_path fname] turns fname into an absolute name, if not
+    already the case, using the default output-directory or the
+    current directory, if the default output-directory is not set *)
```
IIUC the current code if given a relative -output-directory will produce a relative path here. We should absolutify the argument or change this comment (probably absolutify, making it robust to `Cd`).

herbelin replied:
Right. Maybe should we make the output directory absolute at the time of parsing the command line then. Otherwise, if we do `Cd`, it will dynamically modify the output directory, which is probably not be what we want.

Unless you have a better idea, I'll do that.

herbelin replied:
Actually, the same question arises for options `-native-output-dir` and `-nI` (native_include_dirs).

herbelin replied:
"Otherwise, if we do Cd"... (sorry, I was repeating what you had said).

herbelin commented:
@coqbot run full ci

herbelin commented:
I fixed in passing an issue with option `-native-output-dir` which was failing badly when provided with an absolute name. I added support for it, but maybe, it should be forbidden, I don't know.

ejgallego commented on lib/flags.mli:
```diff
@@ -96,3 +96,6 @@ val default_inline_level : int
 val profile_ltac_cutoff : float ref
+
+(** Default output directory *)
+val output_directory : CUnix.physical_path option ref
```
IMO this should be in `system.ml` (use point), not here.

herbelin replied:
Why not. I will consider it.

ejgallego commented on sysinit/coqargs.ml:
```diff
@@ -382,6 +384,10 @@ let parse_args ~usage ~init arglist : t * string list =
       { oval with config = { oval.config with native_output_dir } }
 
+    |"-output-dir" | "-output-directory" ->
+      let dir = CUnix.canonical_dir (next ()) in
```
I think that messing with the directory the user provides is not a good idea. Why are you doing this?

For example, error messages won't contain the original directory, and the extra resolution may cause other issues.

IMHO this is a blocker for this PR.

herbelin replied:
The motivation is to be robust over `Cd`: it is more intuitive that a relative directory given on the command line is resolved relatively to the command line than dynamically depending on which `Cd`s have done so far.

But the question of reporting an error is a good question. So maybe, what can be done is to kept the pair of both the relative output directory and of the directory in which the command has been launched??

ejgallego commented on boot/util.ml:
```diff
@@ -48,7 +48,7 @@ let use_suffix prefix suffix =
   else Filename.concat prefix suffix
 
-let canonical_path_name p =
+let canonical_dir p =
```
Please don't do these kind of changes in unrelated PRs.

Due to the renaming extra noise, it took me much longer to review this PR, and I couldn't do as much of as precise job.

herbelin replied:
Sorry about that, I should have told that the renaming is in a separate commit.

More generally, even with the renaming, I'm not fully satisfied by the terminology used in `CUnix`. We should fix some naming conventions at some time.

SkySkimmer commented:
@herbelin do you want to do some code changes in response to @ejgallego comments or is this ready?

ejgallego commented:
@SkySkimmer I think maybe we should avoid the `canonical_dir` call, and instead respect the original path given (and preprend `pwd` if needed), that can be done in a new commit so the change is contained if you want.

Other than I think the PR is OK.

SkySkimmer commented:
If someone does `coqc -output-dir foo bar.v` and `bar.v` contains
~~~coq
Cd "baz".
Redirect "file" Check True.
~~~
what should happen? The canonical_dir call IIUC is to make it write to `$original_cwd/foo/file.out` instead of `$original_cwd/baz/foo/file.out`.

ejgallego commented:
> The canonical_dir call IIUC is to make it write to `$original_cwd/foo/file.out` instead of `$original_cwd/baz/foo/file.out`.

The canonical_dir call is used to make the directory absolute, but it does a lot more stuff, for example will resolve hard and soft links. That second behavior is the problematic one, as user passes `-output-dir foo`.

If `foo` a symlink, then it will be resolved; that can break other stuff, as people often assume particular layouts which are not invariant under this (actually ill-defined) "path canonization".

What I propose is not to mess with `output-dir` argument except by preprending `pwd` if it is not absolute.

[Note that we must handle absolute stuff for windows paths too, etc...]

herbelin commented:
> What I propose is not to mess with output-dir argument except by preprending pwd if it is not absolute.

We could do that as well.

herbelin commented:
I now moved the `cUnix.ml` things in #18732 and simply prepended the current directory when the output directory is relative.

SkySkimmer commented on lib/system.ml:
```diff
@@ -225,6 +225,21 @@ let is_in_system_path filename =
 
+let warn_using_current_directory =
+  CWarnings.create ~name:"default-output-directory" ~category:CWarnings.CoreCategories.filesystem
+    (fun s ->
+       Pp.(strbrk "Output directory is unset, using \"" ++ str s ++ str "\"." ++ spc () ++
+           strbrk "Use command line option \"-output-directory to set a default directory."))
```
usualy I do
```suggestion
    Pp.(fun s ->
       strbrk "Output directory is unset, using \"" ++ str s ++ str "\"." ++ spc () ++
       strbrk "Use command line option \"-output-directory to set a default directory.")
```
less parentheses that way
(I didn't check indentation in the `` ```suggestion ``)

herbelin replied:
Ok, done.

SkySkimmer commented:
@coqbot merge now
